aiohttp/web_fileresponse.py

```diff
@@ -177,13 +180,23 @@ def _get_file_path_stat_encoding(
         return file_path, file_path.stat(), None
 
     async def prepare(self, request: "BaseRequest") -> Optional[AbstractStreamWriter]:
-        loop = asyncio.get_event_loop()
         # Encoding comparisons should be case-insensitive
         # https://www.rfc-editor.org/rfc/rfc9110#section-8.4.1
         accept_encoding = request.headers.get(hdrs.ACCEPT_ENCODING, "").lower()
-        file_path, st, file_encoding = await loop.run_in_executor(
-            None, self._get_file_path_stat_encoding, accept_encoding
-        )
+
+        loop = asyncio.get_running_loop()
+        try:
+            file_path, st, file_encoding = await loop.run_in_executor(
+                None, self._get_file_path_stat_encoding, accept_encoding
+            )
+        except FileNotFoundError:
+            self.set_status(HTTPNotFound.status_code)
+            return await super().prepare(request)
+
+        # Forbid special files like sockets, pipes, devices, etc.
+        if not S_ISREG(st.st_mode):
+            self.set_status(HTTPForbidden.status_code)
```

Comment:
This seems like a breaking change to not return HTTPNotFound since https://github.com/aio-libs/aiohttp/pull/8538/files#diff-9632858c8fdd45d817126b2735f763b9a2e24fb2800122eb2cb0c9a3b89c85c1L693 previously did

Reply:
I was going to comment myself here.  There are conflicting things written in the old test, which was not working properly.

403 seems like the correct response to me, so I just picked that expecting discussion.

Reply:
403 does seems like a better response. I’m leaning towards keeping it not found though as 

- It’s revealing more information about the filesystem. Need to think about the security implications of that
- It’s technically a breaking change.


Maybe that’s ok at this point since we are going from 3.9 to 3.10. Need to sleep on this

Reply:
It's not escaping the sandbox so I do t think there is a security concern.  Let's keep it as you have it but explicitly call it out in the change message.

Reply:
Spec does permit 404 for stuff that exists but otherwise inaccessible for the "hidden" effect.  However, if that's going to be the `aiohttp` mantra, then it should be consistent (i.e. all permission errors should also give 404).  That's not the case right now which is another reason I used 403 here.

Reply:
It leaks information about something existing on the filesystem.

The only issue I can think of that leading to, is if the app supports uploading of files, using the same name for the upload could replace that file or result in the permissions changing. But, that seems like a bit of a stretch.

From the user perspective, if they are a general user of the app, then a 403 is not useful information and potentially confusing as it suggests they have the wrong login credentials or something. It would only be somewhat useful if the user is also the developer of the app (but, maybe a warning log could resolve this in a safer way).

Reply:
For posterity, here's the short [RFC 9110 section on 403](https://www.rfc-editor.org/rfc/rfc9110#name-403-forbidden).

> It leaks information about something existing on the filesystem.

Right, but if your server becomes vulnerable by that info alone, you've got bigger problems, for which a 404 isn't the solution.

> From the user perspective, if they are a general user of the app, then a 403 is not useful information and potentially confusing as it suggests they have the wrong login credentials or something. 

That case is actually where it seems useful to me.  It says you've done everything right, but the admin didn't give your login the right permissions (as opposed to a 404 which says the app is broken for everyone).

My only strong opinion here would be consistency - either always send 404 instead of 403, or use 403 whenever it exists.  Don't mix and match.  If there's an outcry for security via obscurity, then it would not be difficult to introduce an app flag for a `hidden_forbidden` mode to swap the handful of places it's raised.

Reply:
I didn't merge this as I'm leaving this open for Sam to respond.

Reply:
I don't mind. I'm on the fence at to which way to go.